### ISSUE TITLE
New Labels for BotLand Issues

### ISSUE THREAD

**Issue by DennisRas** [Adam13531/BotLand#47]
So I've added [some new labels](https://github.com/Adam13531/BotLand/labels) and this issue can be used to discuss the labels.  
Should some be removed, renamed or added? Colors? Feedback appreciated.

Here's a text version:

> #EE1E1E - Red
#EE801E - Orange
#EEE21E - Yellow
#17C117 - Green
#1E59EE - Blue
#1EB1EE - Blue light
#9D1EEE - Purple
#EE1EDD - Pink
#EDEDED - Grey

> #EE1E1E - Type: Bug
#9D1EEE - Type: Maintenance
#9D1EEE - Type: Feature
#9D1EEE - Type: Improvement
#1EB1EE - Type: Question
#1EB1EE - Type: Discussion
#EDEDED - Type: Duplicate
#EDEDED - Type: Invalid

> #17C117 - Priority: Low
#EEE21E - Priority: Medium
#EE801E - Priority: High
#EE1E1E - Priority: Critical 

> #1EB1EE - Status: Info Needed
#17C117 - Status: Confirmed
#EEE21E - Status: On hold for now
#17C117 - Status: Review Needed
#EE1EDD - Status: Help Wanted
#EDEDED - Status: Cannot Reproduce
#EDEDED - Status: Rejected

> #1E59EE - Component: Core
#1E59EE - Component: Research
#1E59EE - Component: UI
#1E59EE - Component: Balance
#1E59EE - Component: Mechanics
#1E59EE - Component: Server
#1E59EE - Component: Data
#1E59EE - Component: Documentation & Wiki
#1E59EE - Component: Website & Social

> #17C117 - Changes: Minor (<30 lines)
#EEE21E - Changes: Medium (<250 lines)
#EE801E - Changes: Major (>250 lines)

**Comment by DennisRas** [comment]
I'm thinking the prefix "Component" is a bit too long. Any other suggestions?  
Func:, Area: or something else...

**Comment by HiDeoo** [comment]
What I'm gonna say below is only my **personal** opinion and feedback.

To add more details, I'm using Github daily for multiple years now, for work & personal projects, from small to large repositories, from 1 to 100+ people.

To summarize, I really don't like any of this. With more details:

* For me, a label is a quick way to identify something either using a few letters, 1 word up to a maximum of 2 words and a color.

At the moment, all new labels are basically 3+ words, it defeat the purpose I expect from a label. Having to read a label with more than 1 word is not something I would expect from a label.

Regarding the colors, the idea for me is the same, use few colors, basic colors so your mind can basically memorize them and 1 color is always 1 thing. At the moment, red is used for 3 labels, dark blue is 10+ labels, light dark blue is 3 labels, light other blue is 2 labels, etc. I can't tell from a quick look what is what. It's impossible.

* Too many labels.

For me, labels should used to quickly tag an issue like it's a "bug", it's "invalid" and not a way to avoid reading the issue by having tags like: "bug" - "Redux" - "Game Reducer" - "Line #43" - "Priority: semi low" - "Changes: major (>250 lines) - "Would require a transpilation" - "Help wanted". At one point, it defeats the purposes of just giving a quick look and getting a general idea.

The other downside of having too many labels is that the label inspector only show portion of them.

![screen shot 2017-02-18 at 01 04 27](https://cloud.githubusercontent.com/assets/494699/23087997/d9b1aa82-f578-11e6-8fc1-dcdd4108a95c.png)

Yeah, I can filter them to add them, but to even see all the labels and make sure there is not a more appropriate label than the one I'm choosing in the 30+ labels, I would need to scroll. For me, this is not usable.

Also, for me, an issue should have at max 2/3 tags, which fits the Github UI, if I had 7 tags, issues will take multiple lines on the list. Not sure if that's really usable.

* I thought one of the main goal of labels was for collaborators to tag issues with labels so Adam can get a quick look at it. Some of this labels are totally not usable by collaborators. I think a label should be used by everyone. It should be self explanatory or it's just to specific.

I have no idea how can any of us could know for example how many lines of code an issue would change.

* I don't get the point of types in every labels. It had text for nothing from my point of view. 

    * `Type: bug` => It's a bug, I get it. "Type:" is extra.
    * `Status: On how for now` => On hold, "Status" & "for now" is extra.
    * `Type: improvement` => It's an improvement.

* Duplicates or similar tag:

`bug`, `type: bug`
`Status: Help Wanted`, `Type: question`

* I'm gonna give an example, Lodash, which is one of the most used package at the moment, it has 0 opened issues and around 2500 closed issues. They managed that with only 9 labels:

    * bug
    * change
    * documentation
    * duplicate
    * enhancement
    * invalid
    * question
    * votes needed
    * wontfix

For me it's a perfect example on how to use labels on Github, it's clear, 1 word, 1 unique color for each label, no confusion, you don't spend 10mn tagging an issue and when looking at it, you know immediatly what the issue is about.

Also, all issues they have only get 2 tags max, makes it very clear in 1 second what it's about:

![screen shot 2017-02-18 at 01 21 27](https://cloud.githubusercontent.com/assets/494699/23087999/df5904d0-f578-11e6-892a-8f6f6a11f779.png)

* Personal usage: I think if we check the log, at the moment I'm kinda the only one tagging issues with labels. I'm not gonna spent 10 minutes tagging an issue if it's "related to a bug regarding the core & balance of the game, is an in-game report, considered as high priority (I'm not Adam but assuming that), that can be reproduced but requires more informations so will be on hold at the moment until more infos are provided. Also, the issue contain a question". I would need to add the following tags:

    * bug
    * Type: bug
    * Component: Balance
    * Component: Core
    * in-game report
    * Priority: high
    * Status: confirmed
    * Status: info needed
    * Status: on hold for now
    * Type: question
    * help wanted

That's a no go for me.

**Comment by DennisRas** [comment]
@HiDeoo Alright. Thanks for the feedback. At least the discussion is off to a good start, and any constructive feedback is good feedback. I'll try and see if I can reply to your comment and I appreciate you took the time to partake.

> For me, a label is a quick way to identify something either using a few letters, 1 word up to a maximum of 2 words and a color.

I don't really have much to say about that, as I think that comes down to personal opinion and project needs. A lot of projects use labels to organize and "label" their issues as a tool to prioritize, collaborate, assign and manage their progress/status. The default labels Github provides lack this kind of opportunity.

> Regarding the colors, the idea for me is the same, use few colors, basic colors so your mind can basically memorize them and 1 color is always 1 thing. At the moment, red is used for 3 labels, dark blue is 10+ labels, light dark blue is 3 labels, light other blue is 2 labels, etc. I can't tell from a quick look what is what. It's impossible.

I don't think I've seen a single project that use colors (besides "bug" and "duplicate") grouped together as indicators. The colors I've suggested all have something in common: Dark blue for "Components" (albeit could be renamed as suggested), Light blue for "discussions/comments/reviews/invitation to comment", and then the basic level/priority indicators for a label category (explained further down), such as high, medium, low etc.

> The other downside of having too many labels is that the label inspector only show portion of them.

Can't really comment on that. There's a short scrollbar and a filter and it took me less than 10 seconds to label this issue by scrolling down to each label category and picking one.

> Also, for me, an issue should have at max 2/3 tags, which fits the Github UI, if I had 7 tags, issues will take multiple lines on the list. Not sure if that's really usable.

There should never be that many labels. The point of the labes having prefixes is that only 1 label per prefix can be on an issue at any time.

> I thought one of the main goal of labels was for collaborators to tag issues with labels so Adam can get a quick look at it. Some of this labels are totally not usable by collaborators. I think a label should be used by everyone. It should be self explanatory or it's just to specific.

Granted, the "Changes" category is totally optional and I forgot to mention that. This was for project management only in case more than 1 coder worked on the project. I can see why it isn't important here, but this label collection is just a template crafted in combination with other projects.

> Personal usage: I think if we check the log, at the moment I'm kinda the only one tagging issues with labels. I'm not gonna spent 10 minutes tagging an issue if it's "related to a bug regarding the core & balance of the game, is an in-game report, considered as high priority (I'm not Adam but assuming that), that can be reproduced but requires more informations so will be on hold at the moment until more infos are provided.

I don't understand how it'd take you 10 minutes to label an issue. You have max 5 labels to add (including 1 being optional from the "Changes" category). A new issue is created - you add the labels that you as a collaborator can add and the project owners/coders will add/change whatever label is left.

This isn't set in stone obviously and it's up to Adam and what not, but I figured I would chip in with what I've gathered from other projects. And then there's also the taste and type of management style to take into consideration. Some people like to label or document more than others. :)

**Comment by HiDeoo** [comment]
I think both our opinions on labels are valid, there is no rules so everything is acceptable so I'm not gonna add to your comments which are perfectly fine. You have gathered from multiple repos and I can do the same from my perspective. I think it's just opinions at this point.

The "10mn" part was obviously an exaggeration but I'm not an native english speaker so sometime wording effects are not my specialty ^^ The idea was from my personal experience, I'm used to choose 1 or 2 tag between a list of 7 to 10 tags easily memorable which takes 10s at most. 

Having to scroll to even see all tags, choose 5 tags from a 30ish list of tags that I can't remember at all at the moment, when the number of issue will raise with Beta and later RC, is not something that I'm **personally** willing to do.

**Comment by DennisRas** [comment]
@HiDeoo I agree. There's no right or wrong. This is just a suggestion for another way of managing labels.

I forgot to mention that the reason I think the "Prefix" labelling method (categorizing the labels) suggested here is something that I think fits the development process of this project. Say Adam wants to spend a day working on the UI. He can then filter all the issues related to the UI and have a go at that. That's the key to the label categories.

Regarding the amount of labels, 6-8 of them could be removed/merged. Removing the label category "Changes" would bring the total count down to 19 descriptive labels in 4 categories: Type, Status, Prio, Comp. Still too many?

**Comment by Adam13531** [comment]
@DennisRas I only skimmed this while I was streaming earlier, so now I've gotten a chance to absorb what you've both said. Most of what I'm typing here is just me thinking aloud.

I have not used GitHub extensively, and I'm not sure that I can accurately predict the needs of Bot Land, especially since there's no telling if it even gets popular (and exactly _how_ popular). I feel like all of these decisions need to made under the assumption that the game is at least popular enough to support itself, otherwise there wouldn't be a need for any of these pages!

I wonder if it's worthwhile to figure out a target audience for this particular issues repository. For example, there's no reason why game-design suggestions couldn't go here, but I wonder if they're better suited for something with a real voting system like reddit or UserVoice. We could just have this issue for game/wiki problems, or we could have it be a free-for-all until it proves to be a problem.

Some thoughts on the proposed tags:

Type: Bug
~~Type: Maintenance~~ - not sure if needed
~~Type: Feature~~ - consolidate these as **enhancement**
~~Type: Improvement~~
Type: Question
~~Type: Discussion~~ - I'd try to steer people toward reddit for this
Type: Duplicate - this and "invalid" sound like statuses to me
Type: Invalid

Priority: Low
Priority: Medium
Priority: High
Priority: Critical

Status: Info Needed
Status: Confirmed
Status: On hold for now
~~Status: Review Needed~~ - probably not needed given that we just have one developer
~~Status: Help Wanted~~ - see above
Status: Cannot Reproduce
~~Status: Rejected~~ - just renaming this to "won't fix"

~~Component: Core~~ - lump this and mechanics into "gameplay"
~~Component: Research~~ - don't think this is needed
Component: UI
Component: Balance
~~Component: Mechanics~~ - see Core
~~Component: Server~~ - this doesn't sound like an important tag
~~Component: Data~~ - this seems like Balance or Documentation to me
Component: Documentation & Wiki
Component: Website & Social - I don't think the web site will be important enough to have bugs filed against it, but "social" could still be an important area

This leaves us with this list of 19 labels (down from 31):
Type: Bug
Type: Question
Type: Enhancement

Priority: Low
Priority: Medium
Priority: High
Priority: Critical

Status: Info Needed
Status: Confirmed
Status: On hold for now
Status: Cannot Reproduce
Status: Duplicate
Status: Invalid
Status: Won't Fix

Component: UI
Component: Balance
Component: Gameplay
Component: Documentation & Wiki
Component: Website & Social

19 still sounds kind of high, but "Component" and "Priority" account for 9 of those. Bot Land has enough distinct areas that I think the "Component" label is a good idea, and "Priority" is probably going to be important.

As for colors, we'll have to see how those work out. I don't really have any opinions on them right now.

**Comment by DennisRas** [comment]
Labels reviewed/shortened and the colors streamlined.
Now there's basically 4 categories of labels and 6 main colors.

I don't know what to think/do about the label "in-game report".  
Is it really needed anymore? If the issue is related to something in-game, there'll be browser/game data in the issue anyway and for now the label is kinda spammy/attached to 90%+ issues.

T: Bug
T: Enhancement
T: Question

P: Low
P: Medium
P: High
P: Critical 

S: Info Needed
S: Confirmed
S: On hold
S: Cannot Reproduce
S: Duplicate
S: Invalid
S: Won't fix

C: UI
C: Balance
C: Gameplay
C: Documentation
C: Website & Social

```
#BD1919 - Red
#E07503 - Orange
#E5CD00 - Yellow
#3D9422 - Green
#1972BD - Blue
#7619BD - Purple
#E1E1E1 - Grey
```

**Comment by HiDeoo** [comment]
`T: bug` is second to last.

![qsqpnzw](https://cloud.githubusercontent.com/assets/494699/23138297/d59fc188-f7a6-11e6-9957-b10df1f1f508.gif)

**Comment by Adam13531** [comment]
I got rid of "in-game report". It's probably never going to be used for filtering, and it's obvious enough because there'll be browser/user information directly in the bug when they're filed that way.

I removed the "T:" from all of the "types" so that it looks like this when you open the dialog:

![image](https://cloud.githubusercontent.com/assets/7192897/23149018/f3d6d348-f79d-11e6-9f25-492a6d01ea7f.png)

I don't think I like it though. For consistency and ordering, we could put "A:" before all of them (for "attribute" I guess?). I wonder if it's better to just drop the letter and colon altogether.
